src/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -153,26 +164,90 @@ rawPageLookup !page !key
       | otherwise
       = LookupEntryNotPresent
 
+-- |
+-- __Time:__ \( \mathcal{O}\left( \log_2 n \right) \)
+-- where \( n \) is the number of entries in the 'RawPage'.
+--
+-- Return the least entry number in the 'RawPage' (if it exists)
+-- which is greater than or equal to the supplied 'SerialisedKey'.
+--
+-- The following law always holds \( \forall \mathtt{key} \mathtt{page} \):
+--
+--   * maybe True (key <=) (getRawPageIndexKey . rawPageIndex page . id   =<< rawPageFindKey page key)
+--
+--   * maybe True (key > ) (getRawPageIndexKey . rawPageIndex page . pred =<< rawPageFindKey page key)
+--
+--   * maybe (maximum (rawPageKeys page) < key) (rawPageFindKey page key)
+rawPageFindKey
+    :: RawPage
+    -> SerialisedKey
+    -> Maybe Word16  -- ^ entry number of first entry greater or equal to the key
+rawPageFindKey !page !key
+  | dirNumKeys == 1 = lookup1
+  | otherwise       = case bisectPageToKey page key of
+    KeyNotInPage   -> Nothing
+    KeyFoundAt _ i -> Just $ fromIntegral i
+
+  where
+    !dirNumKeys = rawPageNumKeys page
+
+    lookup1
+      | key <= rawPageKeyAt page 0 = Just 0
+      | otherwise                  = Nothing
+
+-- | This data-type facilitates the code reuse of 'bisectPageToKey' between the
+-- and 'rawPageLookup' and 'rawPageFindKey'.
+data BinarySearchResult
+  = KeyNotInPage
+  | KeyFoundAt !Bool {-# UNPACK #-} !Int
```

Comment:
The Bool here is not explained and generally I think the constructor names are a bit misleading. `KeyNotInPage` means specifically that the key is larger than the largest key in the page. `KeyFoundAt False i` means that the key was *not* found, but the next-largest key can be found at index `i`.

Maybe three constructors would be clearer? (I think the allocation should still be optimised away, but might be worth checking)

```haskell
  = KeyIsTooLarge
  | KeyNotFoundNextIs {-# UNPACK #-} !Int
  | KeyFoundAt {-# UNPACK #-} !Int
```

We can also leave it at two constructors, but then tweaking the name and commenting generously would be nice.

Reply:
I expanded the sum type to be more explicit.

```haskell
data BinarySearchResult                                                                                                         
  = KeyNotFoundExceedsPage                                                                                                      
  | KeyNotFoundNearestIsAt {-# UNPACK #-} !Int                                                                                  
  | KeyWasFoundAt {-# UNPACK #-} !Int 
```

Reply:
`Next` would be a bit clearer than `Nearest`. The previous key might be very "near", but we still return the next one.